documentation/benchmark.md

```diff
@@ -0,0 +1,130 @@
+    * [Benchmark application usage](#31-benchmark-application-usage)
+    * [Adding custom timestamps](#32-adding-custom-timestamps)
+    * [Analyzing results](#33-analyzing-results)
+4. [Limitations and further steps](#4-limitations-and-further-steps)
+
+
+## 1. Introduction
+Due to its nature, OP-TEE is being a solution spanning over several
+architectural layers, where each layer includes its own complex parts.
+For further optimizations of performance, there is a need of tool which will
+provide detailed and precise profiling information for each layer.
+
+It is necessary to receive measurements of latency for:
+* The roundtrip time for going from a client application in normal world,
+down to a Trusted Application and back again.
+* Detailed information for amount of time taken to go through each layer:
+	* libTEEC -> OP-TEE KMOD
+	* OP-TEE KMOD -> OP-TEE OS Core
+	* OP-TEE OS Core -> TA entry point (**not supported yet**)
+	* The same way back
+---
+## 2. Implementation details
+### 2.1 Design overview
+
+Benchmark framework consists of such components:
+1. **Benchmark CA**: a dedicated client application, which is responsible
+for allocating timestamp circular buffers, registering these buffers in
+the **Benchmark PTA**, and consuming all timestamp data, generated
+by all OP-TEE layers. Finally, it puts timestamp data into appropriate
+file with `.ts` extension. Additional details can be found here
+here [optee_benchmark]
+2. **Benchmark PTA**: pseudo TA, which owns all per-cpu circular buffers.
+**Benchmark PTA** must be invoked (by a CA) to register the time stamp
+circular buffers. In turn, the pTA invokes the optee linux driver
+(through some RPC mean) to register this circular buffers in the linux
+kernel layer.
+3. **libTEEC** and **Linux OP-TEE kernel driver** include means to store
+timestamps in the circular buffers once these are registered by the
+**Banchmark pTA** into the "Linux OP-TEE kernel driver".
+
+Benchmark is enabled, all OP-TEE layers (libteec, kernel driver, core)
+do fill the registered timestamp circular buffer with timestamp data for
+all invocation requests
+
+![design_overview][design_overview]
+
+### 2.2 Timestamp source
+
+ARM Performance Monitor Units are used as the main source of timestamp values.
+The reason why this technology was chosen is that it's supported on all
+ARMv7/ARMv8 cores. Besides it can provide precise pre-cpu cycle counter values,
+it's possible to enable EL0 access to all events, so usermode applications can
+directly read cpu counter values from coprocessor registers,
+achieving minimal latency by avoiding additional syscalls to EL1 core.
+
+Besides CPU cycle counter values, timestamp by itself contains also
+information about:
+* Executing CPU core index
+* OP-TEE layer id, where this timestamp was
+obtained from
+* Program counter value when timestamp was logged, which can be used for
+getting a symbol name (a filename and line number)
+
+### 2.3 Call sequence diagram
+
```

Comment:
I agree, but this information is already available in the **2.1 design overview** section